BPF Checks failure seems legit:

```
host_proxy.c:26:1: error: use of undeclared identifier '__config_enable_conntrack_accounting'; did you mean '__check_enable_conntrack_accounting'?
   26 | ASSIGN_CONFIG(bool, enable_conntrack_accounting, true)
```

Do you mind taking a look @atykhyy @julianwiedmann ? :pray:

--- a/bpf/bpf_host.c
+++ b/bpf/bpf_host.c
@@ -296,8 +296,8 @@ handle_ipv6_cont(struct __ctx_buff *ctx, __u32 secctx, const bool from_host,
 		if (from_host) {
 			bool is_host_id = from_host_raw & FROM_HOST_FLAG_HOST_ID;
 
-			ret = __ipv6_host_policy_egress(ctx, is_host_id, ip6, ct_buffer, &trace,
-							ext_err);
+			ret = __ipv6_host_policy_egress(ctx, is_host_id, false, ip6, ct_buffer,
+							&trace, ext_err);
 		} else {
 			ret = __ipv6_host_policy_ingress(ctx, ip6, ct_buffer, &remote_id, &trace,
 							 ext_err);
@@ -518,7 +518,7 @@ int tail_handle_ipv6_from_netdev(struct __ctx_buff *ctx)
 
 # ifdef ENABLE_HOST_FIREWALL
 static __always_inline int
-handle_to_netdev_ipv6(struct __ctx_buff *ctx, __u32 src_sec_identity,
+handle_to_netdev_ipv6(struct __ctx_buff *ctx, bool is_from_proxy, __u32 src_sec_identity,
 		      struct trace_ctx *trace, __s8 *ext_err)
 {
 	void *data, *data_end;
@@ -554,7 +554,8 @@ handle_to_netdev_ipv6(struct __ctx_buff *ctx, __u32 src_sec_identity,
 	srcid = resolve_srcid_ipv6(ctx, ip6, src_sec_identity, &ipcache_srcid);
 
 	/* to-netdev is attached to the egress path of the native device. */
-	return ipv6_host_policy_egress(ctx, srcid, ipcache_srcid, ip6, trace, ext_err);
+	return ipv6_host_policy_egress(ctx, is_from_proxy, srcid, ipcache_srcid, ip6,
+				       trace, ext_err);
 }
 #endif /* ENABLE_HOST_FIREWALL */
 #endif /* ENABLE_IPV6 */
@@ -722,8 +723,8 @@ handle_ipv4_cont(struct __ctx_buff *ctx, __u32 secctx, const bool from_host,
 		if (from_host) {
 			bool is_host_id = from_host_raw & FROM_HOST_FLAG_HOST_ID;
 
-			ret = __ipv4_host_policy_egress(ctx, is_host_id, ip4, ct_buffer, &trace,
-							ext_err);
+			ret = __ipv4_host_policy_egress(ctx, is_host_id, false, ip4, ct_buffer,
+							&trace, ext_err);
 		} else {
 			ret = __ipv4_host_policy_ingress(ctx, ip4, ct_buffer, &remote_id, &trace,
 							 ext_err);
@@ -983,7 +984,7 @@ int tail_handle_ipv4_from_netdev(struct __ctx_buff *ctx)
 
 #ifdef ENABLE_HOST_FIREWALL
 static __always_inline int
-handle_to_netdev_ipv4(struct __ctx_buff *ctx, __u32 src_sec_identity,
+handle_to_netdev_ipv4(struct __ctx_buff *ctx, bool is_from_proxy, __u32 src_sec_identity,
 		      struct trace_ctx *trace, __s8 *ext_err)
 {
 	void *data, *data_end;
@@ -1006,7 +1007,8 @@ handle_to_netdev_ipv4(struct __ctx_buff *ctx, __u32 src_sec_identity,
 	/* We need to pass the srcid from ipcache to host firewall. See
 	 * comment in ipv4_host_policy_egress() for details.
 	 */
-	return ipv4_host_policy_egress(ctx, src_id, ipcache_srcid, ip4, trace, ext_err);
+	return ipv4_host_policy_egress(ctx, is_from_proxy, src_id, ipcache_srcid, ip4,
+				       trace, ext_err);
 }
 #endif /* ENABLE_HOST_FIREWALL */
 #endif /* ENABLE_IPV4 */
@@ -1407,14 +1409,14 @@ int cil_to_netdev(struct __ctx_buff *ctx)
 # endif
 # ifdef ENABLE_IPV6
 	case bpf_htons(ETH_P_IPV6):
-		ret = handle_to_netdev_ipv6(ctx, src_sec_identity,
-					    &trace, &ext_err);
+		ret = handle_to_netdev_ipv6(ctx, magic == MARK_MAGIC_PROXY_EGRESS,
+					    src_sec_identity, &trace, &ext_err);
 		break;
 # endif
 # ifdef ENABLE_IPV4
 	case bpf_htons(ETH_P_IP): {
-		ret = handle_to_netdev_ipv4(ctx, src_sec_identity,
-					    &trace, &ext_err);
+		ret = handle_to_netdev_ipv4(ctx, magic == MARK_MAGIC_PROXY_EGRESS,
+					    src_sec_identity, &trace, &ext_err);
 		break;
 	}
 # endif
@@ -1828,7 +1830,7 @@ from_host_to_lxc(struct __ctx_buff *ctx, __be16 proto, __s8 *ext_err)
 		if (!revalidate_data(ctx, &data, &data_end, &ip6))
 			return DROP_INVALID;
 
-		ret = ipv6_host_policy_egress(ctx, HOST_ID, 0, ip6, &trace, ext_err);
+		ret = ipv6_host_policy_egress(ctx, false, HOST_ID, 0, ip6, &trace, ext_err);
 		break;
 # endif
 # ifdef ENABLE_IPV4
@@ -1842,7 +1844,7 @@ from_host_to_lxc(struct __ctx_buff *ctx, __be16 proto, __s8 *ext_err)
 		 * src_id is HOST_ID. Therefore, we don't need to pass a value
 		 * for the last parameter. That avoids an ipcache lookup.
 		 */
-		ret = ipv4_host_policy_egress(ctx, HOST_ID, 0, ip4, &trace, ext_err);
+		ret = ipv4_host_policy_egress(ctx, false, HOST_ID, 0, ip4, &trace, ext_err);
 		break;
 # endif
 	default:

--- a/bpf/bpf_lxc.c
+++ b/bpf/bpf_lxc.c
@@ -541,7 +541,7 @@ int NAME(struct __ctx_buff *ctx)						\
 	ct_state = (struct ct_state *)&ct_buffer.ct_state;			\
 	tuple = (struct ipv4_ct_tuple *)&ct_buffer.tuple;			\
 										\
-	if (!revalidate_data(ctx, &data, &data_end, &ip4))			\
+	if (!revalidate_data_pull(ctx, &data, &data_end, &ip4))		\
 		return drop_for_direction(ctx, DIR, DROP_INVALID, ext_err);	\
 										\
 	tuple->nexthdr = ip4->protocol;						\
@@ -611,7 +611,7 @@ int NAME(struct __ctx_buff *ctx)						\
 	ct_state = (struct ct_state *)&ct_buffer.ct_state;			\
 	tuple = (struct ipv6_ct_tuple *)&ct_buffer.tuple;			\
 										\
-	if (!revalidate_data(ctx, &data, &data_end, &ip6))			\
+	if (!revalidate_data_pull(ctx, &data, &data_end, &ip6))		\
 		return drop_for_direction(ctx, DIR, DROP_INVALID, ext_err);	\
 										\
 	tuple->nexthdr = ip6->nexthdr;						\

--- a/bpf/lib/host_firewall.h
+++ b/bpf/lib/host_firewall.h
@@ -18,6 +18,34 @@
 #include "proxy.h"
 #include "trace.h"
 
+static __always_inline bool
+need_apply_host_policy_egress(bool is_host_id, bool is_from_proxy)
+{
+	/* Some pod-originating traffic may have a host IP as source IP
+	 * (eg. non-transparent proxy connection, or when using iptables masquerading).
+	 * The response packet will therefore have a host IP as the destination IP.
+	 *
+	 * We don't want to apply egress policy for such packets. But
+	 * to avoid enforcing host policies for response packets to pods, we
+	 * need to create a CT entry for the forward, SNATed packet from the
+	 * pod. Response packets will thus match this CT entry and bypass host
+	 * policies.
+	 * We know the packet is a SNATed packet if the srcid from ipcache is
+	 * HOST_ID, but the actual srcid (derived from the packet mark) isn't.
+	 */
+	if (!is_host_id)
+		return false;
+
+	/* Host packets sent by proxy in response to host requests have had host policy
+	 * already applied to them. We don't want to apply host policy to them again,
+	 * as it creates redirect loops and prevents responses from reaching the proxy.
+	 */
+	if (is_from_proxy)
+		return false;
+
+	return true;
+}
+
 # ifdef ENABLE_IPV6
 static __always_inline bool
 ipv6_host_policy_egress_lookup(struct __ctx_buff *ctx, __u32 src_sec_identity,
@@ -54,7 +82,7 @@ ipv6_host_policy_egress_lookup(struct __ctx_buff *ctx, __u32 src_sec_identity,
 }
 
 static __always_inline int
-__ipv6_host_policy_egress(struct __ctx_buff *ctx, bool is_host_id,
+__ipv6_host_policy_egress(struct __ctx_buff *ctx, bool is_host_id, bool is_from_proxy,
 			  struct ipv6hdr *ip6, struct ct_buffer6 *ct_buffer,
 			  struct trace_ctx *trace, __s8 *ext_err)
 {
@@ -67,6 +95,7 @@ __ipv6_host_policy_egress(struct __ctx_buff *ctx, bool is_host_id,
 	__u32 dst_sec_identity = 0;
 	__u16 proxy_port = 0;
 	__u32 cookie = 0;
+	bool apply_policy;
 
 	trace->monitor = ct_buffer->monitor;
 	trace->reason = (enum trace_reason)ret;
@@ -75,7 +104,8 @@ __ipv6_host_policy_egress(struct __ctx_buff *ctx, bool is_host_id,
 	if (ret == CT_REPLY || ret == CT_RELATED)
 		return CTX_ACT_OK;
 
-	if (is_host_id) {
+	apply_policy = need_apply_host_policy_egress(is_host_id, is_from_proxy);
+	if (apply_policy) {
 		const struct remote_endpoint_info *info;
 		__u32 tunnel_endpoint = 0;
 
@@ -118,7 +148,7 @@ __ipv6_host_policy_egress(struct __ctx_buff *ctx, bool is_host_id,
 			return ret;
 	}
 
-	if (is_host_id) {
+	if (apply_policy) {
 		/* Emit verdict if drop or if allow for CT_NEW. */
 		if (verdict != CTX_ACT_OK || ret != CT_ESTABLISHED)
 			send_policy_verdict_notify(ctx, dst_sec_identity, tuple->dport,
@@ -139,7 +169,7 @@ __ipv6_host_policy_egress(struct __ctx_buff *ctx, bool is_host_id,
 }
 
 static __always_inline int
-ipv6_host_policy_egress(struct __ctx_buff *ctx, __u32 src_id,
+ipv6_host_policy_egress(struct __ctx_buff *ctx,  bool is_from_proxy, __u32 src_id,
 			__u32 ipcache_srcid, struct ipv6hdr *ip6,
 			struct trace_ctx *trace, __s8 *ext_err)
 {
@@ -150,7 +180,7 @@ ipv6_host_policy_egress(struct __ctx_buff *ctx, __u32 src_id,
 	if (ct_buffer.ret < 0)
 		return ct_buffer.ret;
 
-	return __ipv6_host_policy_egress(ctx, src_id == HOST_ID,
+	return __ipv6_host_policy_egress(ctx, src_id == HOST_ID, is_from_proxy,
 					ip6, &ct_buffer, trace, ext_err);
 }
 
@@ -325,7 +355,7 @@ ipv4_host_policy_egress_lookup(struct __ctx_buff *ctx, __u32 src_sec_identity,
 }
 
 static __always_inline int
-__ipv4_host_policy_egress(struct __ctx_buff *ctx, bool is_host_id,
+__ipv4_host_policy_egress(struct __ctx_buff *ctx, bool is_host_id, bool is_from_proxy,
 			  struct iphdr *ip4, struct ct_buffer4 *ct_buffer,
 			  struct trace_ctx *trace, __s8 *ext_err)
 {
@@ -338,6 +368,7 @@ __ipv4_host_policy_egress(struct __ctx_buff *ctx, bool is_host_id,
 	__u32 dst_sec_identity = 0;
 	__u16 proxy_port = 0;
 	__u32 cookie = 0;
+	bool apply_policy;
 
 	trace->monitor = ct_buffer->monitor;
 	trace->reason = (enum trace_reason)ret;
@@ -346,20 +377,8 @@ __ipv4_host_policy_egress(struct __ctx_buff *ctx, bool is_host_id,
 	if (ret == CT_REPLY || ret == CT_RELATED)
 		return CTX_ACT_OK;
 
-	/* Some pod-originating traffic may have a host IP as source IP
-	 * (eg. non-transparent proxy connection, or when using iptables masquerading).
-	 * The response packet will therefore have a host IP as the destination IP.
-	 *
-	 * We don't want to apply egress policy for such packets. But
-	 * to avoid enforcing host policies for response packets to pods, we
-	 * need to create a CT entry for the forward, SNATed packet from the
-	 * pod. Response packets will thus match this CT entry and bypass host
-	 * policies.
-	 * We know the packet is a SNATed packet if the srcid from ipcache is
-	 * HOST_ID, but the actual srcid (derived from the packet mark) isn't.
-	 */
-
-	if (is_host_id) {
+	apply_policy = need_apply_host_policy_egress(is_host_id, is_from_proxy);
+	if (apply_policy) {
 		const struct remote_endpoint_info *info;
 		__u32 tunnel_endpoint = 0;
 
@@ -402,7 +421,7 @@ __ipv4_host_policy_egress(struct __ctx_buff *ctx, bool is_host_id,
 			return ret;
 	}
 
-	if (is_host_id) {
+	if (apply_policy) {
 		/* Emit verdict if drop or if allow for CT_NEW. */
 		if (verdict != CTX_ACT_OK || ret != CT_ESTABLISHED)
 			send_policy_verdict_notify(ctx, dst_sec_identity, tuple->dport,
@@ -423,7 +442,7 @@ __ipv4_host_policy_egress(struct __ctx_buff *ctx, bool is_host_id,
 }
 
 static __always_inline int
-ipv4_host_policy_egress(struct __ctx_buff *ctx, __u32 src_id,
+ipv4_host_policy_egress(struct __ctx_buff *ctx, bool is_from_proxy, __u32 src_id,
 			__u32 ipcache_srcid, struct iphdr *ip4,
 			struct trace_ctx *trace, __s8 *ext_err)
 {
@@ -434,7 +453,8 @@ ipv4_host_policy_egress(struct __ctx_buff *ctx, __u32 src_id,
 	if (ct_buffer.ret < 0)
 		return ct_buffer.ret;
 
-	return __ipv4_host_policy_egress(ctx, src_id == HOST_ID, ip4, &ct_buffer, trace, ext_err);
+	return __ipv4_host_policy_egress(ctx, src_id == HOST_ID, is_from_proxy,
+					ip4, &ct_buffer, trace, ext_err);
 }
 
 static __always_inline bool

--- a/pkg/logging/klog_bridge.go
+++ b/pkg/logging/klog_bridge.go
@@ -19,9 +19,9 @@ import (
 
 var klogErrorOverrides = []logLevelOverride{
 	{
-		// TODO: We can drop the misspelled case here once client-go version is bumped to include:
-		//	https://github.com/kubernetes/client-go/commit/ae43527480ee9d8750fbcde3d403363873fd3d89
-		matcher:     regexp.MustCompile("Failed to update lock (optimitically|optimistically).*falling back to slow path"),
+		// TODO: We can drop this once bumped to new client-go version which has this at info level:
+		// https://github.com/kubernetes/client-go/commit/ea7a7e7cf9697850f17631f79ef4ef45b95c449e.
+		matcher:     regexp.MustCompile("Failed to update.*falling back to slow path"),
 		targetLevel: slog.LevelInfo,
 	},
 }

